src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,571 @@
+        if (sm.find(parsePos.index) && em.find(parsePos.index)) {
+            // get em to the last
+            var c = em.start();
+            while (em.find()) {
+                c = em.start();
+            }
+            em.find(c);
+            var startEnd = sm.end();
+            var endStart = em.start();
+            if (startEnd <= endStart) {
+                var mid = source.substring(startEnd, endStart);
+                var count = mid.split(middleBetween).length + 2;
+                parsed = new MessageFormat(createMessageFormatString(count), locale).parseObject(source, parsePos);
+            }
+        }
+
+        if (parsed == null) {
+            // now try exact number patterns
+            parsed = new MessageFormat(patterns[TWO], locale).parseObject(source, parsePos);
+            if (parsed == null && !patterns[THREE].isEmpty()) {
+                parsed = new MessageFormat(patterns[THREE], locale).parseObject(source, parsePos);
+            }
+        }
+
+        // return the entire source from parsePos if still no match
+        if (parsed == null) {
+            parsed = new String[]{source.substring(parsePos.index)};
+            parsePos.setIndex(source.length());
+        }
+
+        if (parsed instanceof Object[] objs) {
+            parsePos.setErrorIndex(-1);
+            return Arrays.asList(objs);
+        }
+        throw new InternalError("MessageFormat.parseObject() should return Object[]");
+    }
+
+    @Override
+    public AttributedCharacterIterator formatToCharacterIterator(Object arguments) {
+        if (arguments instanceof List<?> objs) {
+            var a = objs.toArray(new Object[0]);
+            return generateMessageFormat(a).formatToCharacterIterator(a);
+        } else {
+            throw new IllegalArgumentException("The arguments should be an Object list");
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj instanceof ListFormat other) {
+            return locale.equals(other.locale) &&
+                Arrays.equals(patterns, other.patterns);
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hash(locale, Arrays.hashCode(patterns));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return
+            """
+            ListFormat [locale: "%s", start: "%s", middle: "%s",  end: "%s", two: "%s", three: "%s"]
+            """.formatted(locale.getDisplayName(), patterns[START], patterns[MIDDLE], patterns[END], patterns[TWO], patterns[THREE]);
+    }
+
+    private MessageFormat generateMessageFormat(Object[] input) {
+        var len = input.length;
+        return switch (len) {
+            case 0 -> throw new IllegalArgumentException("There should at least be one input string");
```

Comment:
It is unfortunate that LDML did not consider the case of an empty list. For example, to supply "NONE" or "n/a" or similar.  The developer will need to deal with that separately.